virttest/libvirt_xml/base.py

```diff
@@ -233,6 +233,57 @@ def virt_xml_validate(filename, schema_name=None):
         cmdresult.stderr = cmdresult.stderr_text
         return cmdresult
 
+    def setup_attrs(self, **attrs):
+        """
+        Setup attributes of an xml object
+
+        :param attrs: dict-type attributes to be set
+
+        Example:
+            dimm_device_attrs = {
+                'mem_model': 'dimm',
+                'target': {'size': 524288, 'size_unit': 'KiB', 'node': 0},
+                'alias': {'name': 'dimm2'},
+                'address': {'attrs': {'type': 'dimm', 'slot': '2', 'base': '0x120000000'}}
+            }
+
+            dimm_device = memory.Memory()
+            dimm_device.setup_attrs(**dimm_device_attrs)
+
+        The content of dimm_device would be:
+            <?xml version='1.0' encoding='UTF-8'?>
+            <memory model="dimm">
+                <target>
+                    <size unit="KiB">524288</size><node>0</node>
+                </target>
+                <alias name="dimm2" />
+                <address base="0x120000000" slot="2" type="dimm" />
+            </memory>
+
+        """
+        reset_all = True if attrs.get('reset_all') is True else False
+
+        for key, value in attrs.items():
+            if key not in self.__all_slots__:
+                raise AttributeError('Cannot set attribute "%s" to %s object.'
+                                     'There is no such attribute.'
+                                     % (key, self.__class__))
+            get_func = eval('self.get_%s' % key)
+
+            # Is XMLElementNest or not
+            subclass = get_func.get('subclass')
+            if subclass is None:
+                setattr(self, key, value)
+            else:
+                # Element tag not found or reset_all
+                if reset_all or not self.xmltreefile.find(key):
+                    subclass_dargs = get_func.get('subclass_dargs')
+                    target_obj = subclass(**subclass_dargs)
```

Comment:
Sorry can not understand here, could you please explain more here with comments? Thanks

Reply:
Ok